### ISSUE TITLE
ANALYZE Results in: ERROR:  invalid byte sequence for encoding "UTF8"

### ISSUE THREAD

**Issue by bsislow** [laurenz/oracle_fdw#348]
When we analyze a foreign table which is very large (billions of rows), we get this error:

`
negval=> analyze mytable;

ERROR:  invalid byte sequence for encoding "UTF8": 0xbf
CONTEXT:  converting column "text_0349" for foreign table scan of "mytable", row 323244
`

We set the sample_percent to 5.

Is there any way to skip these bad rows?

If we don't use the ANALYZE, the performance of the PG query is abysmal; it does a FTS on the table above.

**Comment by laurenz** [comment]
Does the error also happen if you run

    SELECT text_0349 FROM mytable;

That is what I would expect.

You should fix row 323244 of the Oracle table, but of course you will say that you cannot do that.
Moreover, if the first corrupted entry is in row 300000, I'd expect that there are many more broken rows in that large table.

Fortunately, the README describes this useful option:

> **sample_percent** (optional, defaults to "100")
>
>  This option only influences ANALYZE processing and can be useful to
>  ANALYZE very large tables in a reasonable time.
>
>  The value must be between 0.000001 and 100 and defines the percentage of
>  Oracle table blocks that will be randomly selected to calculate PostgreSQL
>  table statistics.  This is accomplished using the `SAMPLE BLOCK (x)`
>  clause in Oracle.

So you could try

    ALTER FOREIGN TABLE mytable OPTIONS (SET sample_percent '0.000001');

and pray that the sample doesn't hit one of the bad rows.

This will also keep Oracle from doing a sequential scan on the table.

**Comment by bsislow** [comment]
Thank you - and yes, we cannot necessarily "fix" that row as this is production data - but - what do you mean by the row is "broken?" What defines "broken" in this context - not readable by PostgreSQL?

Also, how can we find these "broken" rows ahead of time?

Thanks.

**Comment by laurenz** [comment]
No, that means that there is data corruption in the Oracle database.

Sadly, that is pretty common, because Oracle doesn't check strings for integrity if client character set and server character set are identical.

0xBF is the encoding of &ldquo;&#191;&rdquo; in Windows-1252, ISO 8859-1 and similar single-byte encodings. In UTF-8 it would be represented by the two bytes 0xC2BF.

Since your PostgreSQL database encoding is UTF-8, oracle_fdw correctly sets `NLS_LANG` to `AL32UTF8`. Now that must also be the Oracle database character set. Right? So Oracle sends the garbage unfiltered to PostgreSQl, which chokes on it.

How did your Oracle database get corrupted?

Simple. Just set client character set equal to server character set and send data that are *not* correct UTF-8. Oracle will gladly store them as they are. In your case, some buggy application must have sent Spanish text in a single-byte encoding.

How can you find the rows? There is a trick that I developed for my migration tool ora_migrator.
Run the following on Oracle:

    SELECT rowid
    FROM bad_table
    WHERE convert(
             convert(
                bad_column,
                'AL16UTF16'
             ),
             (SELECT value
              FROM nls_database_parameters
              WHERE parameter = 'NLS_CHARACTERSET'
             ),
             'AL16UTF16'
          ) <> bad_column;

An explanation for that black magic: I convert the column from the original encoding to another and back. If everything is fine, the result is the same. If there are corrupt data, Oracle cannot convert them, so it replaces them with &ldquo;replacement characters&rdquo; (rather than throwing an error as it should).

All in all, Oracle is pretty bad when it comes to data integrity.

**Comment by ravisql09** [comment]
Hi Laurenz,

The above query failing for the cases like **multi bite** and **0xe0** records in oracle. Any alternative to eliminate manual work.

Thanks

**Comment by ravisql09** [comment]
By applying TRIM function ,col+''  and above query,I fixed all issues as of now.

Thanks
